### ISSUE TITLE
Locale cannot be set in JiT mode

### ISSUE THREAD

**Issue by sebek64** [angular/angular#16960]
**I'm submitting a ...**  (check one with "x")
```
[X] bug report
```
**Current behavior**

Setting LOCALE_ID via providers passed to bootstrapModule doesn't work. Locale is en-US. There is a hack that aimed to fix this issue, but it is not working (application_module.ts - provider of LOCALE_ID). It is not working, because the factory doesn't get LOCALE_ID injected, so it cannot forward it.

This problem is only in JiT mode. AoT mode works well.

**Expected behavior**

Overriding LOCALE_ID should work in JiT mode.

**Minimal reproduction of the problem with instructions**

http://plnkr.co/edit/4FK0VfJzJtrf5uzEDNfa?p=preview

**What is the motivation / use case for changing the behavior?**

This usage is described in official i18n guide.

**Please tell us about your environment:**

* **Angular version:** 4.1.3
* **Browser:** all | Chrome
* **Language:** TypeScript 2.*

**Comment by ocombe** [comment]
It seems that LOCALE_ID is defined in the main module that imports BrowserModule, if you define it at bootstrap it doesn't work, which is very weird.
In the mean time, you can define this provider in your main module, and it should work for you.

**Comment by ocombe** [comment]
Ok it works if you change your code from `platformBrowserDynamic().bootstrapModule(AppModule, {providers});` to `platformBrowserDynamic(providers).bootstrapModule(AppModule, {providers});`

For some reason, the locale_id provider isn't available in the injector otherwise. I'll keep investigating.

**Comment by hermagrini** [comment]
Hi @ocombe, before I opened a new issue I wanted to check if it was related to this one: Is there anyway so that we can set the locale in JIT via `ngc`'s `--locale` flag? As it currently stands, if we want to be able to set the locale using the flag we need to use `--aot` which makes development builds really slow.

**Comment by ocombe** [comment]
The `--locale` parameter only works for AOT in the cli, sorry. The reason is that for aot the cli rewrites the "main" bootstrap file, and it can add parameters such as the locale. For JIT that's not the case.
Aot development should go faster once the new aot watch mode is finished (the team works on it right now)

**Comment by hermagrini** [comment]
> Aot development should go faster once the new aot watch mode is finished 
that's great to know! thanks for the reply, is there a github issue where I can track it's progress? in any case I'll have to workaround it by rewriting the main bootstrap file as well.
thanks again!

**Comment by ocombe** [comment]
I'm not sure, https://github.com/angular/angular/issues/12867 seems to be outdated.
@tbosch is there an issue/PR to follow for the watch mode ?

**Comment by tbosch** [comment]
@hermagrini watch mode will ship in Angular 5.

**Comment by ocombe** [comment]
Closing this because you no longer need to define the locale at both bootstrap and module level. You should only define it in the main module for JIT.
The cli now has the watch mode for AOT.
The cli supports --locale for JIT & AOT to load locale data required for i18n pipes, but you still need to define it in the main module for JIT. This will not change as we no longer work on supporting JIT in the cli when there is a workaround.
